singularity-eos/base/fast-math/logs.hpp

```diff
@@ -67,11 +67,53 @@
  * we can go into and out of "grid space" reliably and quickly, we're
  * happy, and this function is an EXACT map into and out of that
  * space.
+ *
+ * See ArXiv:2206.08957
  */
 
 namespace singularity {
 namespace FastMath {
 
+/*
+// Integer Aliased versions. Likely not as portable
+template<typename T>
+KOKKOS_FORCEINLINE_FUNCTION
+auto sgn(const T &x) {
+  return (T(0) < x) - (x < T(0));
+}
+
+KOKKOS_FORCEINLINE_FUNCTION
+auto as_int(double f) {
+  return *reinterpret_cast<long long int*>(&f);
+}
+KOKKOS_FORCEINLINE_FUNCTION
+auto as_double(long long int i) {
+  return *reinterpret_cast<double*>(&i);
+}
+
+KOKKOS_FORCEINLINE_FUNCTION
+double fastlg(const double x) {
+  // Magic numbers constexpr because C++ doesn't constexpr reinterpret casts
+  // these are floating point numbers as reinterpreted as integers.
+  // as_int(1.0)
+  constexpr long long int one_as_int = 4607182418800017408;
+  // 1./static_cast<double>(as_int(2.0) - as_int(1.0))
+  constexpr double scale_down = 2.22044604925031e-16;
+  return static_cast<double>(as_int(x) - one_as_int) * scale_down;
+}
+
+KOKKOS_FORCEINLINE_FUNCTION
+double fastpow2(const double x) {
+  // Magic numbers constexpr because C++ doesn't constexpr reinterpret casts
+  // these are floating point numbers as reinterpreted as integers.
+  // as_int(1.0)
+  constexpr long long int one_as_int = 4607182418800017408;
+  // as_int(2.0) - as_int(1.0)
+  constexpr double scale_up = 4503599627370496;
+  return as_double(static_cast<long long int>(x*scale_up) + one_as_int);
+}
+*/
```

Comment:
Will these functions be uncommented later?

Reply:
I second this question. It is best practice not to include commented code unless there is a good reason to include it.

Reply:
The integer aliasing code is faster---but I didn't feel confident uncommenting it now, until I have a better understanding of implications on all systems we care about. I'd also need to add single-precision overloads if needed, as the magic numbers for single-precision are different.

I can remove it again if y'all prefer.

Reply:
I see. Interesting that it's faster -- what about uncommenting these functions but requiring that they be explicitly enabled by a compile time flag? `HIGH_RISK_MATH=ON` or something. That would also make it easier to test on platforms of interest. Could also for now add a check that forces failure if running in single precision mode.

Reply:
why `KOKKOS_` and not `PORTABILITY_` ?

Reply:
> I see. Interesting that it's faster -- what about uncommenting these functions but requiring that they be explicitly enabled by a compile time flag? `HIGH_RISK_MATH=ON` or something. That would also make it easier to test on platforms of interest. Could also for now add a check that forces failure if running in single precision mode.

I can do that.

> why KOKKOS_ and not PORTABILITY_ ?

Because I screwed up when copy-pasting. I'll fix it.

Reply:
Adding `USE_HIGH_RISK_MATH` option and changing from `KOKKOS_` to `PORTABLE_`.